### PR TITLE
Add string::join. Ref: #191

### PR DESCRIPTION
**Pull request by szymonlipinski**
## Motivation and Context

Adds string.join method to the compiler. It supports calling as:

```
separator.join(items)
```

where:

* separator is a string
* items is either ListOf(str) or TupleOf(str)

## How Has This Been Tested?

There are new tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: nativepython/tests/string_compilation_test.py [hunk 1]
```diff
@@ -457,3 +457,67 @@ def test_find_perf(f, s, subs):
...
+        # test longer separator
+        res = f("---", ListOf(str)(["1", "2", "3"]))
+        self.assertEqual("1---2---3", res)
+
+        # test longer items
+        res = f("---", ListOf(str)(["aaa", "bb", "c"]))
+        self.assertEqual("aaa---bb---c", res)
+
+        # test empty separator
+        res = f("", ListOf(str)(["1", "2", "3"]))
```
Comment (braxtonmckee) [inline nativepython/tests/string_compilation_test.py]: need also to test some empty strings in the list itself

File: nativepython/tests/string_compilation_test.py [hunk 2]
```diff
@@ -457,3 +457,67 @@ def test_find_perf(f, s, subs):
...
+        with self.assertRaisesRegex(TypeError, ""):
+            f_tup_int(",", ListOf(int)([1, 2, 3]))
+
+        # test passing list of other types than strings
+        @Compiled
+        def f_int(sep: str, items: ListOf(int)) -> str:
+            return sep.join(items)
+
+        with self.assertRaisesRegex(TypeError, ""):
+            f_int(",", ListOf(int)([1, 2, 3]))
```
Comment (braxtonmckee) [inline nativepython/tests/string_compilation_test.py:561]: we should check that this also works on a Dict(str, str) or a ConstDict(str, str), since both of those are currently iterable, and we need to verify that theywork.

File: typed_python/StringType.cpp [hunk 3]
```diff
@@ -814,3 +814,96 @@ void StringType::assign(instance_ptr self, instance_ptr other) {
...
+
+    StringType::Make()->constructor((instance_ptr)outString);
+
+    // return empty string when there is nothing to join
+    if(toJoin->count == 0)
+    {
+        return;
+    }
+
+    static ListOfType *inputList = ListOfType::Make(StringType::Make());
```
Comment (braxtonmckee) [inline typed_python/StringType.cpp]: this variable should be called `listOfStrType`. OTherwise it looks like a data variable.

File: nativepython/tests/string_compilation_test.py [hunk 4]
```diff
@@ -457,3 +457,67 @@ def test_find_perf(f, s, subs):
                 .format(c_elapsed, elapsed)
         )
         """
+
+    def check_string_join(self, f):
+        """This function contains tests for implementation of the string join
+        for both: ListOf(str) and TupleOf(str).
+        The tests for the types are as separate functions, so it will be easier
+        to find the failing line later.
+        """
```
Comment (braxtonmckee) [inline nativepython/tests/string_compilation_test.py]: need some tests with strings that are not just 1 byte per codepoint. Try joining with two-byte and four-byte codepoints in both separator and individual string values.

**Review by braxtonmckee — CHANGES_REQUESTED** [review]
looks good. I requested a couple of variable renames and some more tests.

File: nativepython/tests/string_compilation_test.py (hunk 1 above)
Comment (szymonlipinski) [reply nativepython/tests/string_compilation_test.py]: done

File: typed_python/StringType.cpp (hunk 3 above)
Comment (szymonlipinski) [reply typed_python/StringType.cpp]: done

File: nativepython/tests/string_compilation_test.py (hunk 2 above)
Comment (szymonlipinski) [reply nativepython/tests/string_compilation_test.py:561]: done

**Comment by szymonlipinski** [comment]
One interesting thing is that the ConstDict has different ordering of the items than the Dict. It looks like Dict stores them in the order passed in constructor, while the ConstDict removed the empty string and ordered the keys.

**Comment by braxtonmckee** [comment]
ConstDict does keep things sorted. But what do you mean it removed the
empty string? That sounds like a bug. Is it a problem with the join or the
ConstDict itself?

File: nativepython/tests/string_compilation_test.py (hunk 4 above)
Comment (szymonlipinski) [reply nativepython/tests/string_compilation_test.py]: done

**Comment by szymonlipinski** [comment]
I think all is fine with the ConstDict.
